core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -162,68 +174,145 @@ public synchronized void respondAfterAsyncAction(boolean forcedRefresh) {
+        private final AtomicBoolean refreshed = new AtomicBoolean(false);
+        private final AtomicReference<Exception> syncFailure = new AtomicReference<>(null);
+        private final RespondingWriteResult respond;
+        private final IndexShard indexShard;
+        private final WriteRequest<?> request;
+        private final ESLogger logger;
+
+        AsyncAfterWriteAction(final IndexShard indexShard,
+                             final WriteRequest<?> request,
+                             @Nullable final Translog.Location location,
+                             final RespondingWriteResult respond,
+                             final ESLogger logger) {
+            this.indexShard = indexShard;
+            this.request = request;
+            boolean waitUntilRefresh = false;
+            switch (request.getRefreshPolicy()) {
+                case IMMEDIATE:
+                    indexShard.refresh("refresh_flag_index");
+                    refreshed.set(true);
+                    break;
+                case WAIT_UNTIL:
+                    if (location != null) {
+                        waitUntilRefresh = true;
+                        pendingOps.incrementAndGet();
+                    }
+                    break;
+                case NONE:
+                    break;
+                default:
+                    throw new IllegalArgumentException("unknown refresh policy: " + request.getRefreshPolicy());
+            }
+            this.waitUntilRefresh = waitUntilRefresh;
+            this.respond = respond;
+            this.location = location;
+            if ((sync = indexShard.getTranslogDurability() == Translog.Durability.REQUEST && location != null)) {
+                pendingOps.incrementAndGet();
+            }
+            this.logger = logger;
+            assert pendingOps.get() >= 0 && pendingOps.get() <= 2 : "pendingOpts was: " + pendingOps.get();
+        }
+
+        /** calls the response listener if all pending operations have returned otherwise it just decrements the pending opts counter.*/
+        private void maybeFinish() {
+            final int numPending = pendingOps.decrementAndGet();
+            if (numPending == 0) {
+                if (syncFailure.get() != null) {
+                    respond.onFailure(syncFailure.get());
+                } else {
+                    respond.onSuccess(refreshed.get());
+                }
+            }
+            assert numPending == 0 || numPending == 1: "numPending must either 1 or 0 but was " + numPending ;
+        }
+
+        void run() {
+            // we either respond immediately ie. if we we don't fsync per request or wait for refresh
+            // OR we got an pass async operations on and wait for them to return to respond.
+            indexShard.maybeFlush();
+            if (pendingOps.get() == 0) {
+                respond.onSuccess(refreshed.get());
+            } else {
```

Comment:
maybe we could increment pendingOps by one more in the constructor and just call `maybeFinish` here instead of this if/else?